tempesta_fw/http.c

```diff
@@ -2021,15 +2021,16 @@ tfw_http_conn_cli_drop(TfwCliConn *cli_conn)
 	 */
 	spin_lock(&cli_conn->seq_qlock);
 	list_for_each_entry_safe(req, tmp, seq_queue, msg.seq_list) {
-		set_bit(TFW_HTTP_B_REQ_DROP, req->flags);
-		list_del_init(&req->msg.seq_list);
 		/*
-		 * Response is processed and removed from fwd_queue, need to be
-		 * destroyed.
+		 * Request must be destroyed if the response is fully processed
+		 * and removed from fwd_queue.
 		 */
-		if (req->resp
-		    && test_bit(TFW_HTTP_B_RESP_READY, req->resp->flags))
-		{
+		bool unused = req->resp
+			&& test_bit(TFW_HTTP_B_RESP_READY, req->resp->flags);
+		set_bit(TFW_HTTP_B_REQ_DROP, req->flags);
+		smp_mb__after_atomic();
```

Comment:
Why the request can not be freed right here and on the next line `req->msg` dereference lead to crash?

Reply:
Agree with @krizhanovsky: race is still possible here. In my opinion, the simple solution (but maybe not the best) is to lock `seq_qlock` on the other side (`tfw_http_resp_cache_cb()` and others) - during `TFW_HTTP_B_REQ_DROP` checking and request removal.
Also, in this context - it seems that we have the similar problem [here](https://github.com/tempesta-tech/tempesta/blob/8e828719bd03c38aa4d2f3b475753bcdf6429302/tempesta_fw/http.c#L1969): verifying `list_empty_careful(&req->msg.seq_list)` before the lock for the queue is taken.

Reply:
Agree with @krizhanovsky: race is still possible here. In my opinion, the simple solution (but maybe not the best) is to lock `seq_qlock` on the other side (`tfw_http_resp_cache_cb()` and others) - during `TFW_HTTP_B_REQ_DROP` checking and request removal.
Also, in this context - it seems that we have the similar problem [here](https://github.com/tempesta-tech/tempesta/blob/8e828719bd03c38aa4d2f3b475753bcdf6429302/tempesta_fw/http.c#L1969): verifying `list_empty_careful(&req->msg.seq_list)` before the lock for the queue is taken.

Reply:
Agree with @krizhanovsky: race is still possible here. In my opinion, the simple solution (but maybe not the best) is to lock `seq_qlock` on the other side (`tfw_http_resp_cache_cb()` and others) - during `TFW_HTTP_B_REQ_DROP` checking and request removal.
Also, in this context - it seems that we have the similar problem [here](https://github.com/tempesta-tech/tempesta/blob/8e828719bd03c38aa4d2f3b475753bcdf6429302/tempesta_fw/http.c#L1969): verifying `list_empty_careful(&req->msg.seq_list)` before the lock for the queue is taken.

Reply:
There is no need to lock seq_queue when `TFW_HTTP_B_REQ_DROP` is checked all over the  the code. 
If `TFW_HTTP_B_REQ_DROP` flag is set, the request is already removed from the seq_queue and can be accessed only by `resp->req` link. The `RESP_READY` flag for response is set under seq_queue lock, so it's guaranteed that the same request is not accessed by both `tfw_http_conn_cli_drop()` and `tfw_http_resp_fwd()`.

In the same time if the response is not ready, the request is not destroyed in `tfw_http_conn_cli_drop()`. When other thread tests the `DROP` flag, it destroys the request. It's safe since it's just performance optimisation. Without the check the response-request pair will be passed into `tfw_http_resp_fwd()` and will be destroyed there.

The only race I have found was in `resp_and_log()` and I've fixed it.

Reply:
OK, makes sense. If there is logical dependency between `req->msg.seq_list` (a request is in the list) and the flag, then don't we need `smp_mb__before_atomic()` between the list deletion and setting the flag? BTW why do we need the `smp_mb__after_atomic()`?

Reply:
Yeah, this solution is better then additional locks.
Now the sequence
```
list_del_init(&req->msg.seq_list);
set_bit(TFW_HTTP_B_REQ_DROP, req->flags);
```
should not be reordered to avoid races, thus `smp_mb__before_atomic()` is desirable between them.

Reply:
Fixed memory barriers.